sql, kv: add sql.mutations.max_row_size guardrails

michae2 opened the pull request:
**kv: set Batch.pErr during Batch.prepare**

If we fail to construct a Batch (e.g. fail to marshal a key or value)
then an error will be placed in the resultsBuf and the batch will not
actually be sent to the layers below. In this case we still need to set
Batch.pErr, so that Batch.MustPErr is able to return a roachpb.Error to
higher layers without panicking.

I imagine in practice we never fail to marshal the key or value, so we
have never seen this panic in the wild.

Release note: None

Release justification: Bug fix.

**sql: add sql.mutations.max_row_size.log guardrail (large row logging)**

Addresses: #67400

Add sql.mutations.max_row_size.log, a new cluster setting which
controls large row logging. Rows larger than this size will have their
primary keys logged to the SQL_PERF or SQL_INTERNAL_PERF channels
whenever the SQL layer puts them into the KV layer.

This logging takes place in rowHelper, which is used by both
row.Inserter and row.Updater. Most of the work is plumbing
settings.Values and SessionData into rowHelper, and adding a new
structured event type.

Release note (ops change): A new cluster setting,
sql.mutations.max_row_size.log, was added, which controls large row
logging. Whenever a row larger than this size is written (or a single
column family if multiple column families are in use) a LargeRow event
is logged to the SQL_PERF channel (or a LargeRowInternal event is logged
to SQL_INTERNAL_PERF if the row was added by an internal query). This
could occur for INSERT, UPSERT, UPDATE, CREATE TABLE AS, CREATE INDEX,
ALTER TABLE, ALTER INDEX, IMPORT, or RESTORE statements. SELECT, DELETE,
TRUNCATE, and DROP are not affected by this setting.

Release justification: Low risk, high benefit change to existing
functionality. This adds logging whenever a large row is written to the
database. Default is 64 MiB, which is also the default for
kv.raft.command.max_size, meaning on a cluster with default settings
statements writing these rows will fail with an error anyway.

**sql: add sql.mutations.max_row_size.err guardrail (large row errors)**

Addresses: #67400

Add sql.mutations.max_row_size.err, a new cluster setting similar to
sql.mutations.max_row_size.log, which limits the size of rows written to
the database. Statements trying to write a row larger than this will
fail with an error. (Internal queries will not fail with an error, but
will log a LargeRowInternal event to the SQL_INTERNAL_PERF channel.)

We're reusing eventpb.CommonLargeRowDetails as the error type, out of
convenience.

Release note (ops change): A new cluster setting,
sql.mutations.max_row_size.err, was added, which limits the size of rows
written to the database (or individual column families, if multiple
column families are in use). Statements trying to write a row larger
than this will fail with a code 54000 (program_limit_exceeded) error.
(Internal queries writing a row larger than this will not fail, but will
log a LargeRowInternal event to the SQL_INTERNAL_PERF channel.) This
limit is enforced for INSERT, UPSERT, and UPDATE statements. CREATE
TABLE AS, CREATE INDEX, ALTER TABLE, ALTER INDEX, IMPORT, and RESTORE
will not fail with an error, but will log LargeRowInternal events to the
SQL_INTERNAL_PERF channel. SELECT, DELETE, TRUNCATE, and DROP are not
affected by this limit. **Note that existing rows violating the limit
*cannot* be updated, unless the update shrinks the size of the row
below the limit, but *can* be selected, deleted, altered, backed-up, and
restored.** For this reason we recommend using the accompanying setting
sql.mutations.max_row_size.log in conjunction with
SELECT pg_column_size() queries to detect and fix any existing large
rows before lowering sql.mutations.max_row_size.err.

Release justification: Low risk, high benefit change to existing
functionality. This causes statements adding large rows to fail with an
error. Default is 512 MiB, which was the maximum KV size in 20.2 as of
https://github.com/cockroachdb/cockroach/pull/61818 and also the default
range_max_bytes in 21.1, meaning rows larger than this were not possible
in 20.2 and are not going to perform well in 21.1.

michae2 committed:
kv: set Batch.pErr during Batch.prepare

If we fail to construct a Batch (e.g. fail to marshal a key or value)
then an error will be placed in the resultsBuf and the batch will not
actually be sent to the layers below. In this case we still need to set
Batch.pErr, so that Batch.MustPErr is able to return a roachpb.Error to
higher layers without panicking.

I imagine in practice we never fail to marshal the key or value, so we
have never seen this panic in the wild.

Release note: None

Release justification: Bug fix.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67953)
<!-- Reviewable:end -->

michae2 commented:
Obviously this still needs testing, but I'm hoping to solicit feedback about whether this is a good idea in the first place.

jordanlewis commented:
cc @andreimatei as well.

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67953)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @michae2, @nvanbenschoten, and @sumeerbhola)

___
*[pkg/kv/batch.go, line 13 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MfNtyra1x3-BhTpCzId:-MfNtyra1x3-BhTpCzIe:blts456) ([raw file](https://github.com/cockroachdb/cockroach/blob/1875bd6e1a6e619614b20e70b8e5d43f81d104f4/pkg/kv/batch.go#L13)):*
> ```Go
> package kv
> 
> import (
> ```

It seems weird to me that the new limit is enforced inside the `Batch`. Isn't there a higher layer that can just as easily enforce it - like the `tableWriter`? If there's not too many distinct callers that want this, I'd rather keep it in the caller(s).
___
*[pkg/sql/tablewriter.go, line 38 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MfNvYCgBs72wJ4sUokx:-MfNvYCgBs72wJ4sUoky:b-jye3wg) ([raw file](https://github.com/cockroachdb/cockroach/blob/1875bd6e1a6e619614b20e70b8e5d43f81d104f4/pkg/sql/tablewriter.go#L38)):*
> ```Go
> 
> var maxKeyValueSize = settings.RegisterByteSizeSetting(
> 	"sql.mutations.kv_max_size",
> ```

is it feasible to talk in terms of some SQL concept - like a col fam / index entry - instead of "key-value entry".


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67953)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @michae2, @nvanbenschoten, and @sumeerbhola)

___
*[pkg/kv/batch.go, line 13 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MfNtyra1x3-BhTpCzId:-MfNx09A8dht0zuIRCST:ba519z8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1875bd6e1a6e619614b20e70b8e5d43f81d104f4/pkg/kv/batch.go#L13)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

It seems weird to me that the new limit is enforced inside the `Batch`. Isn't there a higher layer that can just as easily enforce it - like the `tableWriter`? If there's not too many distinct callers that want this, I'd rather keep it in the caller(s).
</blockquote></details>

After reading though how the errors make it out of the batch in a deferred way, I want this logic to live out of the batch even more.
___
*[pkg/kv/batch.go, line 385 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MfNwGhGBFLOLLEm5tlU:-MfNwGhHBHD6GYN9s1bF:b-tqyj7a) ([raw file](https://github.com/cockroachdb/cockroach/blob/1875bd6e1a6e619614b20e70b8e5d43f81d104f4/pkg/kv/batch.go#L385)):*
> ```Go
> func (b *Batch) checkKVSize(k roachpb.Key, v roachpb.Value) error {
> 	if size := len(k) + len(v.RawBytes); b.MaxKVBytes != 0 && size > b.MaxKVBytes {
> 		return fmt.Errorf(
> ```

use`WithCandidateCode`. And generally prefer `errors.Errorf` to `fmt.Errorf` except in cases where you're sure you'll never want a stack.


<!-- Sent from Reviewable.io -->

bdarnell commented:
How exactly does this relate to `kv.raft.command.max_size`? It looks like this is meant to be a slightly "softer" limit but it's not clear what exactly is allowed that would be blocked by the existing limit. In particular the fact that updates are not allowed unless they shrink the row back below the limit appears to be the same behavior as the existing limit, and this is a problem for the customer who motivated this PR. 

If we do have both a SQL and KV-level setting for batch size limits, I think it would make sense for the SQL one to have its default no higher than the KV-level default (probably by setting the SQL-level default to the same as the current KV limit and bumping the KV limit up a bit).

andreimatei commented:
> How exactly does this relate to kv.raft.command.max_size?

This one operates at the level of single column families (right?).

> If we do have both a SQL and KV-level setting for batch size limits, I think it would make sense for the SQL one to have its default no higher than the KV-level default (probably by setting the SQL-level default to the same as the current KV limit and bumping the KV limit up a bit).

What I'd like clarified in comments and perhaps in the validator is the relationship with the new `sql.mutations.mutation_batch_byte_size`

nvb reviewed (commented):
> This one operates at the level of single column families (right?).

Right, the limit here operates at the level of a single key-value. `kv.raft.command.max_size` operates at the level of an entire Raft proposal, which may include multiple KVs. `kv.raft.command.max_size` may also block writes that look smaller than they actually end up being, due to the size of their corresponding intent. @dt recently found that it's quite difficult to predict the size of a KV write ahead of time, because we include the transaction key in intents, so even a 10 byte `Put` may be 5MB large if it's transaction has a large archor key.

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-:-MfYzwdVBHbkPcZDpbKS:b-5owgc0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @dt, @michae2, @nvanbenschoten, and @sumeerbhola)

___
*[pkg/kv/batch.go, line 13 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MfNtyra1x3-BhTpCzId:-MfZ0EQV7E72k3FmyrSC:b2tdlji) ([raw file](https://github.com/cockroachdb/cockroach/blob/1875bd6e1a6e619614b20e70b8e5d43f81d104f4/pkg/kv/batch.go#L13)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

After reading though how the errors make it out of the batch in a deferred way, I want this logic to live out of the batch even more.
</blockquote></details>

After reading through this, I was left with the same thought. I don't think this needs to live in the KV batch. And so if there is only a few callers, we may be better off lifting this up a level.


<!-- Sent from Reviewable.io -->

michae2 commented:
Thanks for the feedback, everyone. Yes, you have it right about the relationship with `kv.raft.command.max_size`. I agree about lifting this out of the KV batch and shall do so.

> In particular the fact that updates are not allowed unless they shrink the row back below the limit appears to be the same behavior as the existing limit, and this is a problem for the customer who motivated this PR.

We are discussing this with the customer. The trouble is that, when not using multiple column families, at this level it is difficult to distinguish between something like:

```
-- before new limit
INSERT INTO table (key, shortpayload, longpayload) VALUES (42, 99, "giant text ... ");
-- after new limit
UPDATE table SET shortpayload = 100 WHERE key = 42;
```

and:

```
-- before new limit
INSERT INTO table VALUES (42, 99, NULL);
-- after new limit
UPDATE table SET longpayload = "giant text ..." WHERE key = 42;
```

If multiple column families are in use, as we recommend for this type of schema, then only the column family with the giant text column will be affected by the size limit, and the first update will succeed.

> If we do have both a SQL and KV-level setting for batch size limits, I think it would make sense for the SQL one to have its default no higher than the KV-level default (probably by setting the SQL-level default to the same as the current KV limit and bumping the KV limit up a bit).

Makes sense to me.

sumeerbhola commented:
Is this PR meant to supersede https://github.com/cockroachdb/cockroach/pull/67537 (that PR came up in https://cockroachlabs.slack.com/archives/C0KB9Q03D/p1626287781175700)?

michae2 commented:
> Is this PR meant to supersede #67537 (that PR came up in https://cockroachlabs.slack.com/archives/C0KB9Q03D/p1626287781175700)?

This PR is similar to, but distinct in purpose from the batch_byte_size setting. The batch_byte_size setting controls the number of put / del operations sent in one batch, using the size of operations as a guide, but does not limit the size of any one operation. The setting introduced in this PR limits the size of any one operation.

More changes to come today, so feel free to hold off reading for a while.

michae2 committed:
sql: add sql.mutations.max_row_size.log guardrail (large row logging)

Addresses: #67400

Add sql.mutations.max_row_size.log, a new cluster setting which
controls large row logging. Rows larger than this size will have their
primary keys logged to the SQL_PERF or SQL_INTERNAL_PERF channels
whenever the SQL layer puts them into the KV layer.

This logging takes place in rowHelper, which is used by both
row.Inserter and row.Updater. Most of the work is plumbing
settings.Values and SessionData into rowHelper, and adding a new
structured event type.

Release note (ops change): A new cluster setting,
sql.mutations.max_row_size.log, was added, which controls large row
logging. Whenever a row larger than this size is written (or a single
column family if multiple column families are in use) a LargeRow event
is logged to the SQL_PERF channel (or a LargeRowInternal event is logged
to SQL_INTERNAL_PERF if the row was added by an internal query). This
could occur for INSERT, UPSERT, UPDATE, CREATE TABLE AS, CREATE INDEX,
ALTER TABLE, ALTER INDEX, IMPORT, or RESTORE statements. SELECT, DELETE,
TRUNCATE, and DROP are not affected by this setting.

Release justification: Low risk, high benefit change to existing
functionality. This adds logging whenever a large row is written to the
database. Default is 64 MiB, which is also the default for
kv.raft.command.max_size, meaning on a cluster with default settings
statements writing these rows will fail with an error anyway.

michae2 reviewed (commented):
- Check has been moved up into SQL `rowHelper`.
- Single setting has been split into two, *warn* and *error*. The latest push only contains the *warn* setting, which causes an new `LargeRow` event to be logged. *error* setting is in-progress. I'm hoping to backport the *warn* setting.
- Default now matches `kv.raft.command.max_size` default.
- Added a test.

The error setting is still in-progress, but this is RFAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-:-MhKXAfS-w7ID4GyPmns:bic7a8h)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @dt, @michae2, @nvanbenschoten, and @sumeerbhola)

___
*[pkg/kv/batch.go, line 13 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MfNtyra1x3-BhTpCzId:-MhKWOP3AopAVpbXejAj:b-yo6m1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/1875bd6e1a6e619614b20e70b8e5d43f81d104f4/pkg/kv/batch.go#L13)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

After reading through this, I was left with the same thought. I don't think this needs to live in the KV batch. And so if there is only a few callers, we may be better off lifting this up a level.
</blockquote></details>

Check has been lifted up into `row.rowHelper`, entirely within SQL layer. I kept the small fix to `Batch` in the first commit.
___
*[pkg/sql/tablewriter.go, line 38 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MfNvYCgBs72wJ4sUokx:-MhKWlG--UMHvP8ARu-n:b-d6dnu6) ([raw file](https://github.com/cockroachdb/cockroach/blob/1875bd6e1a6e619614b20e70b8e5d43f81d104f4/pkg/sql/tablewriter.go#L38)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

is it feasible to talk in terms of some SQL concept - like a col fam / index entry - instead of "key-value entry".
</blockquote></details>

I decided to simply use "row".


<!-- Sent from Reviewable.io -->

michae2 committed:
sql: add sql.mutations.max_row_size.err guardrail (large row errors)

Addresses: #67400

Add sql.mutations.max_row_size.err, a new cluster setting similar to
sql.mutations.max_row_size.log, which limits the size of rows written to
the database. Statements trying to write a row larger than this will
fail with an error. (Internal queries will not fail with an error, but
will log a LargeRowInternal event to the SQL_INTERNAL_PERF channel.)

We're reusing eventpb.CommonLargeRowDetails as the error type, out of
convenience.

Release note (ops change): A new cluster setting,
sql.mutations.max_row_size.err, was added, which limits the size of rows
written to the database (or individual column families, if multiple
column families are in use). Statements trying to write a row larger
than this will fail with a code 54000 (program_limit_exceeded) error.
(Internal queries writing a row larger than this will not fail, but will
log a LargeRowInternal event to the SQL_INTERNAL_PERF channel.) This
limit is enforced for INSERT, UPSERT, and UPDATE statements. CREATE
TABLE AS, CREATE INDEX, ALTER TABLE, ALTER INDEX, IMPORT, and RESTORE
will not fail with an error, but will log LargeRowInternal events to the
SQL_INTERNAL_PERF channel. SELECT, DELETE, TRUNCATE, and DROP are not
affected by this limit. **Note that existing rows violating the limit
*cannot* be updated, unless the update shrinks the size of the row
below the limit, but *can* be selected, deleted, altered, backed-up, and
restored.** For this reason we recommend using the accompanying setting
sql.mutations.max_row_size.log in conjunction with
SELECT pg_column_size() queries to detect and fix any existing large
rows before lowering sql.mutations.max_row_size.err.

Release justification: Low risk, high benefit change to existing
functionality. This causes statements adding large rows to fail with an
error. Default is 512 MiB, which was the maximum KV size in 20.2 as of
https://github.com/cockroachdb/cockroach/pull/61818 and also the default
range_max_bytes in 21.1, meaning rows larger than this were not possible
in 20.2 and are not going to perform well in 21.1.

rytaft reviewed (commented):
This is looking great! I'll let others approve since I'm not so familiar with all these code paths.

Is there a reason you don't think you'll be able to backport the error setting (assuming the default is that the error is disabled)?

> Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 2 of 2 files at r3, 5 of 5 files at r4, 24 of 24 files at r5, 2 of 2 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-:-MhPjjzqDX14PhzHdCq-:b-dcty7h)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @dt, @michae2, and @sumeerbhola)

___
*[pkg/sql/slow_query_log_test.go, line 25 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhPixbl8P-_0Bz2XMxU:-MhPixbl8P-_0Bz2XMxV:boj34kc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d07e0d3d91536e38c16d6fe6c9b3b94103f90af/pkg/sql/slow_query_log_test.go#L25)):*
> ```Go
> 	errRe    string
> 	logRe    string
> 	expected int
> ```

what does expected mean here? (add a comment)
___
*[pkg/sql/slow_query_log_test.go, line 213 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhPjGNR0DT34FtPPOVb:-MhPjGNR0DT34FtPPOVc:bizb2lb) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d07e0d3d91536e38c16d6fe6c9b3b94103f90af/pkg/sql/slow_query_log_test.go#L213)):*
> ```Go
> }
> 
> func TestSlowQueryLog(t *testing.T) {
> ```

nit: I'd give this test (and filename) a different name since it's confusing that large row events are also tested here. Maybe you could also just add this test to event_log_test.go?
___
*[pkg/sql/row/helper.go, line 43 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhPVCmj2Akm-TA9ejSG:-MhPVCmk7lwn5OLjJLk4:b-ng01gp) ([raw file](https://github.com/cockroachdb/cockroach/blob/7af6ba3dee5044d5e0725cd0b9c36668504d2595/pkg/sql/row/helper.go#L43)):*
> ```Go
> 	"sql.mutations.max_row_size.warn",
> 	"maximum size of row (or column family if multiple column families are in use) that SQL can "+
> 		"write to the KV store, above which an event is logged to SQL_PERF or SQL_INTERNAL_PERF "+
> ```

why would it be written to one or the other? Maybe add "depending on whether or not the transaction is internal"?
___
*[pkg/sql/sessiondatapb/session_data.proto, line 77 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhPSyFD8g8G5vAbaCa4:-MhPSyFE5mVhQfO_TQFV:bvvlmmu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a20dff0a042a52ea8dc1bc66a13008021df6dd6f/pkg/sql/sessiondatapb/session_data.proto#L77)):*
> ```protobuf
>                                               (gogoproto.stdduration) = true];
>   // Internal indicates whether this query came from InternalExecutor or an
>   // internal planner.
> ```

nit: sounds like these are the only two options. Perhaps you could say:
```
  // Internal is true if this query came from InternalExecutor or an
  // internal planner.
```
Is that correct?


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Latest push should fix the CI errors, as well as address comments (thank you!). Also adds sql.mutations.max_row_size.err. Testcases for that are coming.

> Is there a reason you don't think you'll be able to backport the error setting (assuming the default is that the error is disabled)?

No particular reason, it's just more consequential, so I'm more cautious. I made the default 512 MiB.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-:-MhQvfoB2z0G3BSXUY5N:bhrurpm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @dt, and @rytaft)

___
*[pkg/kv/batch.go, line 385 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MfNwGhGBFLOLLEm5tlU:-MhQr3gb1TBF2eWAkj7G:b-ndjt1h) ([raw file](https://github.com/cockroachdb/cockroach/blob/1875bd6e1a6e619614b20e70b8e5d43f81d104f4/pkg/kv/batch.go#L385)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

use`WithCandidateCode`. And generally prefer `errors.Errorf` to `fmt.Errorf` except in cases where you're sure you'll never want a stack.
</blockquote></details>

Thanks, done. Hopefully I've managed this leaf error thing correctly. :slightly_smiling_face:
___
*[pkg/sql/row/helper.go, line 43 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhPVCmj2Akm-TA9ejSG:-MhQvRZB8t0N-nDl_WnX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7af6ba3dee5044d5e0725cd0b9c36668504d2595/pkg/sql/row/helper.go#L43)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

why would it be written to one or the other? Maybe add "depending on whether or not the transaction is internal"?
</blockquote></details>

Done.
___
*[pkg/sql/sessiondatapb/session_data.proto, line 77 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhPSyFD8g8G5vAbaCa4:-MhQvVhy383AxK9nusfi:bspsrmj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a20dff0a042a52ea8dc1bc66a13008021df6dd6f/pkg/sql/sessiondatapb/session_data.proto#L77)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: sounds like these are the only two options. Perhaps you could say:
```
  // Internal is true if this query came from InternalExecutor or an
  // internal planner.
```
Is that correct?
</blockquote></details>

Oh, good point. Done.
___
*[pkg/sql/slow_query_log_test.go, line 25 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhPixbl8P-_0Bz2XMxU:-MhQvYmmBBKnfZf_4zlY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d07e0d3d91536e38c16d6fe6c9b3b94103f90af/pkg/sql/slow_query_log_test.go#L25)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

what does expected mean here? (add a comment)
</blockquote></details>

Done.
___
*[pkg/sql/slow_query_log_test.go, line 213 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhPjGNR0DT34FtPPOVb:-MhQvZVSD5tLffxSZaoy:b6lmqrd) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d07e0d3d91536e38c16d6fe6c9b3b94103f90af/pkg/sql/slow_query_log_test.go#L213)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: I'd give this test (and filename) a different name since it's confusing that large row events are also tested here. Maybe you could also just add this test to event_log_test.go?
</blockquote></details>

Ah, I didn't know about event_log_test, thank you!


<!-- Sent from Reviewable.io -->

michae2 commented:
Added a few tests for `sql.mutations.max_row_size.err`, and made some linters happy.

rytaft reviewed (commented):
> Reviewed 5 of 5 files at r13, 17 of 27 files at r14, 1 of 2 files at r15, 9 of 9 files at r16.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67953)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @dt, @michae2, and @rytaft)

___
*[pkg/sql/row/helper.go, line 236 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhUJev1BPbt4fdd1pJk:-MhUJev1BPbt4fdd1pJl:bhgj7t6) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d07e0d3d91536e38c16d6fe6c9b3b94103f90af/pkg/sql/row/helper.go#L236)):*
> ```Go
> 		details := eventpb.CommonLargeRowDetails{
> 			RowSize:    size,
> 			MaxRowSize: rh.maxRowSizeWarn,
> ```

Might still help to know what limit was violated (especially due to the new logic you added for internal queries)
___
*[pkg/sql/row/inserter.go, line 182 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhUUzR979d0tQoK7DkK:-MhUUzR979d0tQoK7DkL:b-v5zoyh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9867863ec2e066fe331abfa48b45d4a7439ae6/pkg/sql/row/inserter.go#L182)):*
> ```Go
> 	for i := range secondaryIndexEntries {
> 		e := &secondaryIndexEntries[i]
> 		if err := ri.Helper.checkRowSize(
> ```

Why do we need to check the row size for secondary indexes? Won't the primary check be sufficient?
___
*[pkg/util/log/eventpb/sql_audit_events.proto, line 116 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhUTlGJ4N3uzIZTer4R:-MhUTlGKAN6qq-bPzXNZ:bueheqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9867863ec2e066fe331abfa48b45d4a7439ae6/pkg/util/log/eventpb/sql_audit_events.proto#L116)):*
> ```protobuf
> // LargeRow is recorded when a statement tries to write a row larger than
> // cluster setting `sql.mutations.max_row_size.log` to the database. Multiple
> // LargeRow events will be recorded for statements writing multiple large rows.
> ```

Could this cause a lot of log spam (especially for internal queries)?


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67953)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @dt, and @rytaft)

___
*[pkg/sql/row/helper.go, line 236 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhUJev1BPbt4fdd1pJk:-MhX9CGMCrUM7Dwji1SF:b5gvona) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d07e0d3d91536e38c16d6fe6c9b3b94103f90af/pkg/sql/row/helper.go#L236)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Might still help to know what limit was violated (especially due to the new logic you added for internal queries)
</blockquote></details>

Good point. This ambiguity only exists for the log events, so I've added `ViolatesMaxRowSizeErr` to the log events, but not to the error message.
___
*[pkg/sql/row/inserter.go, line 182 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhUUzR979d0tQoK7DkK:-MhX9WHj4iOF1CgEV_kF:b68vw2s) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9867863ec2e066fe331abfa48b45d4a7439ae6/pkg/sql/row/inserter.go#L182)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Why do we need to check the row size for secondary indexes? Won't the primary check be sufficient?
</blockquote></details>

Oh, great point! I removed the secondary index checks, which simplified many things.
___
*[pkg/util/log/eventpb/sql_audit_events.proto, line 116 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhUTlGJ4N3uzIZTer4R:-MhX9cvhBa9542PsFT43:b-4bwn5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9867863ec2e066fe331abfa48b45d4a7439ae6/pkg/util/log/eventpb/sql_audit_events.proto#L116)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Could this cause a lot of log spam (especially for internal queries)?
</blockquote></details>

Short answer: the logging can be disabled. Right now it defaults to 64 MiB for logging and 512 MiB for error. This seemed high enough to prevent spam, as 64 MiB is the default for the raft command size limit.

Long answer: it depends on how many large rows the cluster has, and what the user considers spam. This was designed with the assumption that the user wants to detect and fix rows larger than `sql.mutations.max_row_size.log` as they enter the system, and so logging the PK is a straightforward way to do that and enables mechanical fixing of large rows. But yes, if it is set inappropriately low it could produce a lot of log messages. In that sense it is very similar to slow query logging (and it is written to the same logging channel).


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> great work!

> Reviewed 5 of 5 files at r18, 17 of 28 files at r19, 1 of 2 files at r20, 10 of 10 files at r21.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-:-MhZHUj174qmCjymr-B9:b-2zonaq)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @dt, and @rytaft)

___
*[pkg/util/log/eventpb/sql_audit_events.proto, line 116 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhUTlGJ4N3uzIZTer4R:-MhZHHF_BKA-uRKA1LKt:b7e4hsb) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9867863ec2e066fe331abfa48b45d4a7439ae6/pkg/util/log/eventpb/sql_audit_events.proto#L116)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Short answer: the logging can be disabled. Right now it defaults to 64 MiB for logging and 512 MiB for error. This seemed high enough to prevent spam, as 64 MiB is the default for the raft command size limit.

Long answer: it depends on how many large rows the cluster has, and what the user considers spam. This was designed with the assumption that the user wants to detect and fix rows larger than `sql.mutations.max_row_size.log` as they enter the system, and so logging the PK is a straightforward way to do that and enables mechanical fixing of large rows. But yes, if it is set inappropriately low it could produce a lot of log messages. In that sense it is very similar to slow query logging (and it is written to the same logging channel).
</blockquote></details>

Makes sense, thanks


<!-- Sent from Reviewable.io -->

michae2 commented:
@andreimatei would you be willing to sign off on the small fix to `kv.Batch` (first commit)?

michae2 commented:
(Plan for backport is to completely disable both settings.)

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67953)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @dt, @michae2, and @rytaft)

___
*[pkg/sql/row/helper.go, line 50 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhZT7cP5cRsO1UCwhmC:-MhZT7cP5cRsO1UCwhmD:bbekmyc) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2cd875e28073d0c7e46b3aec0655a5a10822fd0/pkg/sql/row/helper.go#L50)):*
> ```Go
> 	func(size int64) error {
> 		if size != 0 && size < maxRowSizeFloor {
> 			return fmt.Errorf(
> ```

In #68707 Raphael switched all `fmt.Errorf` into `errors.Newf`.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
- Switched `fmt.Errorf` to `errors.Newf`.
- Added 0 setting to `sql.mutations.max_row_size.err` to disable.
- Cleaned up internal / shouldErr logic in `checkRowSize` (thanks @yuzefovich!).
- Squashed into 3 commits and added release justifications.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-:-MhoDTNp99XBFuv-SFJL:b3u3nlb)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @dt, @rytaft, and @yuzefovich)

___
*[pkg/sql/row/helper.go, line 50 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhZT7cP5cRsO1UCwhmC:-MhoDfBm60bVRedZf9rE:b-y3101r) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2cd875e28073d0c7e46b3aec0655a5a10822fd0/pkg/sql/row/helper.go#L50)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

In #68707 Raphael switched all `fmt.Errorf` into `errors.Newf`.
</blockquote></details>

Oh, thanks! Switched.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 23 of 31 files at r23, 10 of 10 files at r24.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-:-MhoUjckD9zq_yZxV1z9:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @dt, @rytaft, and @yuzefovich)



<!-- Sent from Reviewable.io -->

michae2 commented:
Added a short test for backup and restore of existing large rows.

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r25.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-:-Mhp2MrA0-6qXyCYOqzN:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @dt, @rytaft, and @yuzefovich)



<!-- Sent from Reviewable.io -->

andreimatei reviewed (approved):
First commit <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo removing an `if b.pErr == nil {`

> Release note (bug fix): Fix an extremely rare panic in Batch.MustPErr.

This message won't make sense to a user. Scrap it.

Out of curiosity, how did run into the need for the first commit? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-:-Mhp4Fhb8mHjtt30nkZ8:b-f96ics)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @andreimatei, @dt, @michae2, @rytaft, and @yuzefovich)

___
*[pkg/kv/batch.go, line 97 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhpB3cGANwALZ4dxlPf:-MhpB3cGANwALZ4dxlPg:b-wa6nvd) ([raw file](https://github.com/cockroachdb/cockroach/blob/419c6639ad4405fc9d293a4627ce4e0fff8726cc/pkg/kv/batch.go#L97)):*
> ```Go
> }
> 
> func (b *Batch) prepare() error {
> ```

Please comment that this method checks for marshalling errors; it took me a while to convince that this method is necessary at all.
And consider renaming it to `validate` or something. It hardly does any "preparing".
___
*[pkg/kv/batch.go, line 97 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhpBmqGEvj32OR5wpXu:-MhpBmqGEvj32OR5wpXv:btigsg4) ([raw file](https://github.com/cockroachdb/cockroach/blob/419c6639ad4405fc9d293a4627ce4e0fff8726cc/pkg/kv/batch.go#L97)):*
> ```Go
> }
> 
> func (b *Batch) prepare() error {
> ```

consider adding a test for this method, given that you're fixing something
___
*[pkg/kv/batch.go, line 99 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-Mhp78V4-vC7HeTN2saV:-Mhp78V5CKVIi9j-EeWb:b-cu5bd3) ([raw file](https://github.com/cockroachdb/cockroach/blob/419c6639ad4405fc9d293a4627ce4e0fff8726cc/pkg/kv/batch.go#L99)):*
> ```Go
> func (b *Batch) prepare() error {
> 	if err := b.resultErr(); err != nil {
> 		if b.pErr == nil {
> ```

Can `b.pErr` be set at this point? I don't think it can, so I think we don't need this condition. Having it is confusing.
Perhaps you've copied this guard from `sendAndFill`, but there it's used to differentiate (I think) between an error result and an RPC error (which is not present in the results). 
___
*[pkg/kv/batch.go, line 100 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhpCLUk32eYPl9epOdc:-MhpCLUk32eYPl9epOdd:b-id95ns) ([raw file](https://github.com/cockroachdb/cockroach/blob/419c6639ad4405fc9d293a4627ce4e0fff8726cc/pkg/kv/batch.go#L100)):*
> ```Go
> 	if err := b.resultErr(); err != nil {
> 		if b.pErr == nil {
> 			b.pErr = roachpb.NewError(err)
> ```

since this pattern of both saving an error and returning it is unusual, consider adding a comment that we're doing this analogously to `sendAndFill`. 
___
*[pkg/kv/batch.go, line 104 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhpAYNU8PXA-wnf0PlA:-MhpAYNU8PXA-wnf0PlB:bqhnmim) ([raw file](https://github.com/cockroachdb/cockroach/blob/419c6639ad4405fc9d293a4627ce4e0fff8726cc/pkg/kv/batch.go#L104)):*
> ```Go
> 		return err
> 	}
> 	return nil
> ```

nit: if you make `err` above not be scoped to the `if`, you can always `return err`.
___
*[pkg/kv/batch.go, line 302 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-Mhp54i919RSzig_Egr4:-Mhp54i919RSzig_Egr5:b-g1x92f) ([raw file](https://github.com/cockroachdb/cockroach/blob/419c6639ad4405fc9d293a4627ce4e0fff8726cc/pkg/kv/batch.go#L302)):*
> ```Go
> // if one exists.
> func (b *Batch) resultErr() error {
> 	for _, r := range b.Results {
> ```

I think this change is not worth taking the blame on this stanza.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Thanks for the reviews!

> Out of curiosity, how did run into the need for the first commit?

The prototype of these guardrails was implemented at the `kv.Batch` level, and used this path to return errors. So I noticed when the errors caused a panic. :slightly_smiling_face:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-:-MhpZmGM5l9h7QoDAl0m:b703d63)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andreimatei, @dt, @rytaft, and @yuzefovich)

___
*[pkg/kv/batch.go, line 97 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhpB3cGANwALZ4dxlPf:-Mhp_KUR4sWd6tK26sgm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/419c6639ad4405fc9d293a4627ce4e0fff8726cc/pkg/kv/batch.go#L97)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Please comment that this method checks for marshalling errors; it took me a while to convince that this method is necessary at all.
And consider renaming it to `validate` or something. It hardly does any "preparing".
</blockquote></details>

Done.
___
*[pkg/kv/batch.go, line 97 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhpBmqGEvj32OR5wpXu:-Mhp_LOo7zyh6grcVqyz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/419c6639ad4405fc9d293a4627ce4e0fff8726cc/pkg/kv/batch.go#L97)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

consider adding a test for this method, given that you're fixing something
</blockquote></details>

Done.
___
*[pkg/kv/batch.go, line 99 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-Mhp78V4-vC7HeTN2saV:-Mhp_MMMBHh9gejmX46E:b92dqp5) ([raw file](https://github.com/cockroachdb/cockroach/blob/419c6639ad4405fc9d293a4627ce4e0fff8726cc/pkg/kv/batch.go#L99)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Can `b.pErr` be set at this point? I don't think it can, so I think we don't need this condition. Having it is confusing.
Perhaps you've copied this guard from `sendAndFill`, but there it's used to differentiate (I think) between an error result and an RPC error (which is not present in the results). 
</blockquote></details>

Agreed (yes I did copy it from `sendAndFill`). Done.
___
*[pkg/kv/batch.go, line 100 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhpCLUk32eYPl9epOdc:-Mhp_EZV6Kn_5aQl0lNM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/419c6639ad4405fc9d293a4627ce4e0fff8726cc/pkg/kv/batch.go#L100)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

since this pattern of both saving an error and returning it is unusual, consider adding a comment that we're doing this analogously to `sendAndFill`. 
</blockquote></details>

Done.
___
*[pkg/kv/batch.go, line 104 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-MhpAYNU8PXA-wnf0PlA:-Mhp_G-D7GRRDER1SZ-y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/419c6639ad4405fc9d293a4627ce4e0fff8726cc/pkg/kv/batch.go#L104)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: if you make `err` above not be scoped to the `if`, you can always `return err`.
</blockquote></details>

Done.
___
*[pkg/kv/batch.go, line 302 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/67953#-Mhp54i919RSzig_Egr4:-Mhp_H-Y6Ce-qs25xDv5:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/419c6639ad4405fc9d293a4627ce4e0fff8726cc/pkg/kv/batch.go#L302)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think this change is not worth taking the blame on this stanza.
</blockquote></details>

Good point, done.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
nit: the PR description should be updated according to the latest commit message (I'm not sure where we're taking the release note from).

michae2 commented:
TFTRs!

bors r=rytaft,andreimatei

michae2 commented:
Note that when this is backported, https://github.com/cockroachdb/cockroach/pull/69457 will also need to be backported.
